mirgecom/viscous.py

```diff
@@ -0,0 +1,419 @@
+        Fluid solution
+
+    Returns
+    -------
+    :class:`~meshmode.dof_array.DOFArray`
+        The maximum stable timestep at each node.
+    """
+    from grudge.dt_utils import characteristic_lengthscales
+    from mirgecom.fluid import compute_wavespeed
+
+    length_scales = characteristic_lengthscales(cv.array_context, discr)
+
+    mu = 0
+    d_alpha_max = 0
+    transport = eos.transport_model()
+    if transport:
+        mu = transport.viscosity(eos, cv)
+        d_alpha_max = \
+            get_local_max_species_diffusivity(
+                cv.array_context, discr,
+                transport.species_diffusivity(eos, cv)
+            )
+
+    return(
+        length_scales / (compute_wavespeed(eos, cv)
+        + ((mu + d_alpha_max) / length_scales))
+    )
+
+
+def get_viscous_cfl(discr, eos, dt, cv):
+    """Calculate and return node-local CFL based on current state and timestep.
+
+    Parameters
+    ----------
+    discr: :class:`grudge.eager.EagerDGDiscretization`
+        the discretization to use
+    eos: :class:`~mirgecom.eos.GasEOS`
+        A gas equation of state
+    dt: float or :class:`~meshmode.dof_array.DOFArray`
+        A constant scalar dt or node-local dt
+    cv: :class:`~mirgecom.fluid.ConservedVars`
+        The fluid conserved variables
+
+    Returns
+    -------
+    :class:`~meshmode.dof_array.DOFArray`
+        The CFL at each node.
+    """
+    return dt / get_viscous_timestep(discr, eos=eos, cv=cv)
+
+
+def get_local_max_species_diffusivity(actx, discr, d_alpha):
+    """Return the maximum species diffusivity at every point.
+
+    Parameters
+    ----------
+    actx: :class:`arraycontext.ArrayContext`
+        Array context to use
+    discr: :class:`grudge.eager.EagerDGDiscretization`
+        the discretization to use
+    d_alpha: numpy.ndarray
+        Species diffusivities
+    """
+    if len(d_alpha) <= 0:
+        return 0
+    if not isinstance(d_alpha[0], DOFArray):
+        return max(d_alpha)
+
+    return_dof = []
+    for i in range(len(d_alpha[0])):
+        stacked_diffusivity = actx.np.stack([x[i] for x in d_alpha])
+
+        n_species, ni1, ni0 = stacked_diffusivity.shape
+
+        @memoize_in(discr, ("max_species_diffusivity", n_species, i))
+        def make_max_kernel():
+            # fun fact: arraycontext needs these exact loop names to work (even
+            # though a loopy kernel can have whatever iterator names the user wants)
+            # TODO: see if the opposite order [i0, i1, i2] is faster due to higher
+            # spatial locality, causing fewer cache misses
+            return arraycontext.make_loopy_program(
+                "{ [i1,i0,i2]: 0<=i1<ni1 and 0<=i0<ni0 and 0<=i2<n_species}",
+                "out[i1,i0] = max(i2, a[i2,i1,i0])"
+            )
+
+        return_dof.append(
+            actx.call_loopy(make_max_kernel(), a=stacked_diffusivity)["out"])
+
+    return DOFArray(actx, tuple(return_dof))
```

Comment:
#492

Reply:
I am really confused why we don't need a kernel here?

Reply:
The kernel is inside `actx.np.maximum`. It computes the elementwise maximum between two arrays. `reduce` calls it successively on each species component, i.e.:

```python
max01 = actx.np.maximum(d_alpha[0], d_alpha[1])
max012 = actx.np.maximum(max01, d_alpha[2])
max0123 = actx.np.maximum(max012, d_alpha[3])
...etc...
```

Reply:
Tricksies. 👍